OpenRA.Mods.Common/Traits/Carryall.cs

```diff
@@ -10,206 +10,314 @@
 	{
-		[Desc("Set to false when the carryall should not automatically get new jobs.")]
-		public readonly bool Automatic = true;
+		public readonly int LoadingDelay = 0;
+		public readonly int UnLoadingDelay = 0;
 
-		public object Create(ActorInitializer init) { return new Carryall(init.Self, this); }
+		[VoiceReference]
+		public readonly string Voice = "Move";
+
+		public virtual object Create(ActorInitializer init) { return new Carryall(init.Self, this); }
 	}
 
-	public class Carryall : INotifyBecomingIdle, INotifyKilled, ISync, IRender, INotifyActorDisposing
+	public class Carryall : INotifyKilled, ISync, IRender, INotifyActorDisposing, IIssueOrder, IResolveOrder, IOrderVoice
 	{
+		public enum CarryallState
+		{
+			Idle,
+			Reserved,
+			Carrying
+		}
+
 		readonly Actor self;
-		readonly WDist carryHeight;
 		readonly CarryallInfo info;
+		readonly AircraftInfo aircraftInfo;
+		readonly IMove move;
+		readonly IFacing facing;
 
 		// The actor we are currently carrying.
-		[Sync] public Actor Carrying { get; internal set; }
-		public bool IsCarrying { get; internal set; }
-
-		// TODO: Use ActorPreviews so that this can support actors with multiple sprites
-		Animation anim;
+		[Sync] public Actor Carryable { get; private set; }
+		public CarryallState State { get; private set; }
 
-		public bool IsBusy { get; internal set; }
+		IActorPreview[] carryablePreview = null;
+		WDist carryableHeight;
 
 		public Carryall(Actor self, CarryallInfo info)
 		{
 			this.self = self;
 			this.info = info;
 
-			IsBusy = false;
-			IsCarrying = false;
+			Carryable = null;
+			State = CarryallState.Idle;
 
-			var helicopter = self.Info.TraitInfoOrDefault<AircraftInfo>();
-			carryHeight = helicopter != null ? helicopter.LandAltitude : WDist.Zero;
+			aircraftInfo = self.Info.TraitInfoOrDefault<AircraftInfo>();
+			move = self.Trait<IMove>();
+			facing = self.Trait<IFacing>();
 		}
 
-		public void OnBecomingIdle(Actor self)
+		void INotifyActorDisposing.Disposing(Actor self)
 		{
-			if (info.Automatic)
-				FindCarryableForTransport();
+			if (State == CarryallState.Carrying)
+			{
+				Carryable.Dispose();
+				Carryable = null;
+			}
 
-			if (!IsBusy)
-				self.QueueActivity(new HeliFlyCircle(self));
+			UnreserveCarryable();
 		}
 
-		// A carryable notifying us that he'd like to be carried
-		public bool RequestTransportNotify(Actor carryable)
+		void INotifyKilled.Killed(Actor self, AttackInfo e)
 		{
-			if (IsBusy || !info.Automatic)
-				return false;
-
-			if (ReserveCarryable(carryable))
+			if (State == CarryallState.Carrying)
 			{
-				self.QueueActivity(false, new PickupUnit(self, carryable));
-				self.QueueActivity(true, new DeliverUnit(self));
-				return true;
+				if (Carryable.IsInWorld && !Carryable.IsDead)
+					Carryable.Kill(e.Attacker);
+				Carryable = null;
 			}
 
+			UnreserveCarryable();
+		}
+
+		public virtual bool RequestTransportNotify(Actor carryable)
```

Comment:
I'm not sure it makes sense to define this here - it probably wants to stay inside `AutoCarryall`.